File: libcraft/generators/python/simplified_block.py
from common import load_minecraft_json, load_feather_json, camel_case, generate_enum, generate_enum_property, output
from re import compile

blocks = load_minecraft_json("blocks.json")
simplified_block = load_feather_json("simplified_block.json")

regexes = {}
for name, regex in simplified_block['regexes'].items():
    regexes[name] = compile(regex)

variants = []
mapping = {}
for name in regexes:
    variants.append(camel_case(name))

for block in blocks:
    name = block['name']
    block_variant = camel_case(name)

    # Detect which SimplifiedBlockKind matches this block.
    found = False
    for simplified, regex in regexes.items():
        if regex.match(name) is not None:
            mapping[block_variant] = "SimplifiedBlockKind::" + camel_case(simplified)
            found = True
            break

    if not found:
        # Default to block variant
        variants.append(block_variant)
        mapping[block_variant] = "SimplifiedBlockKind::" + block_variant

output_data = "use crate::BlockKind;" + generate_enum("SimplifiedBlockKind", variants)
output_data += generate_enum_property("BlockKind", "simplified_kind", "SimplifiedBlockKind", mapping)
output("blocks/src/simplified_block.rs", output_data)

File: data_generators/src/generators/simplified_block.rs
use convert_case::{Case, Casing};
use indexmap::map::IndexMap;
use regex::Regex;

use crate::utils::*;

pub fn generate() {
    let simplified_blocks: SimplifiedBlocks = load_libcraft_json("simplified_block.json").unwrap();
    let blocks: Vec<IdAndName> = load_minecraft_json("blocks.json").unwrap();

    let mut out = quote! { use crate::BlockKind; };

    let mappings = blocks
        .into_iter()
        .map(|block| {
            (
                block.name.to_case(Case::UpperCamel),
                simplified_blocks
                    .regexes
                    .iter()
                    .filter(|(_, regexp)| Regex::new(regexp).unwrap().is_match(&block.name))
                    .map(|(kind, _)| kind.to_case(Case::UpperCamel))
                    .next()
                    .unwrap_or_else(|| block.name.to_case(Case::UpperCamel)),
            )
        })
        .collect::<IndexMap<_, _>>();

    let mut variants = mappings.values().collect::<Vec<_>>();
    variants.sort();
    variants.dedup();
    out.extend(generate_enum!(SimplifiedBlockKind, variants));
    out.extend(generate_enum_property!(
        BlockKind,
        "simplified_kind",
        SimplifiedBlockKind,
        mappings
            .into_iter()
            .map(|(key, value)| (key, {
                let kind = format_ident!("{}", value);
                quote! { SimplifiedBlockKind::#kind }
            }))
            .collect(),
    ));

    output(
        "libcraft/blocks/src/simplified_block.rs",
        out.to_string().as_str(),
    );
}

#[derive(Deserialize)]
struct SimplifiedBlocks {
    regexes: IndexMap<String, String>,
}
